File: frontend/src/api/displaySettings.ts
import apiClient, { type ApiError } from './client';

/** Кейс §5.9/§5.11: проектные настройки отображения (гость + перенос в файле). */
export interface ProjectDisplaySettingsPayload {
  heatcalc?: Record<string, unknown>;
  electrical?: Record<string, unknown>;
  specification?: Record<string, unknown>;
}

export interface ProjectDisplaySettings {
  project_id: string;
  version: number;
  settings: ProjectDisplaySettingsPayload;
}

export const PROJECT_DISPLAY_SETTINGS_QUERY_KEY = 'project-display-settings';

export async function getProjectDisplaySettings(
  projectId: string,
): Promise<ProjectDisplaySettings> {
  const { data } = await apiClient.get<ProjectDisplaySettings>(
    `/projects/${projectId}/display-settings`,
  );
  return data;
}

export async function updateProjectDisplaySettings(
  projectId: string,
  expectedVersion: number,
  settings: ProjectDisplaySettingsPayload,
): Promise<ProjectDisplaySettings> {
  const { data } = await apiClient.put<ProjectDisplaySettings>(
    `/projects/${projectId}/display-settings`,
    { expected_version: expectedVersion, settings },
  );
  return data;
}

export function isDisplaySettingsVersionConflict(error: unknown): boolean {
  const apiError = error as ApiError | null;
  return apiError?.status === 409
    && apiError?.code === 'PROJECT_DISPLAY_SETTINGS_VERSION_CONFLICT';
}

File: backend/app/api/v1/display_settings.py
"""Project display settings endpoints (кейс §5.9 / §5.11)."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import CurrentPrincipal, require_any
from app.schemas.project_display_settings import (
    ProjectDisplaySettingsResponse,
    ProjectDisplaySettingsUpdateRequest,
)
from app.services.project_display_settings_service import (
    ProjectDisplaySettingsConflictError,
    ProjectDisplaySettingsService,
    ProjectDisplaySettingsTooLargeError,
)
from app.services.project_service import ProjectAccessError, ProjectNotFoundError

router = APIRouter()
_require_any = require_any()


@router.get(
    "/{project_id}/display-settings",
    response_model=ProjectDisplaySettingsResponse,
    summary="Get project display settings",
)
async def get_project_display_settings(
    project_id: UUID,
    principal: CurrentPrincipal = Depends(_require_any),
    db: AsyncSession = Depends(get_db),
) -> ProjectDisplaySettingsResponse:
    try:
        return await ProjectDisplaySettingsService(db).get(project_id, principal)
    except ProjectNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ProjectAccessError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc


@router.put(
    "/{project_id}/display-settings",
    response_model=ProjectDisplaySettingsResponse,
    summary="Update project display settings (optimistic version)",
)
async def update_project_display_settings(
    project_id: UUID,
    data: ProjectDisplaySettingsUpdateRequest,
    principal: CurrentPrincipal = Depends(_require_any),
    db: AsyncSession = Depends(get_db),
) -> ProjectDisplaySettingsResponse:
    try:
        return await ProjectDisplaySettingsService(db).update(project_id, data, principal)
    except ProjectDisplaySettingsConflictError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "PROJECT_DISPLAY_SETTINGS_VERSION_CONFLICT",
                "message": str(exc),
                "details": {
                    "expected_version": exc.expected_version,
                    "current_version": exc.current_version,
                },
            },
        ) from exc
    except ProjectDisplaySettingsTooLargeError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
            detail={
                "code": "PROJECT_DISPLAY_SETTINGS_TOO_LARGE",
                "message": str(exc),
                "details": {
                    "size_bytes": exc.size_bytes,
                    "limit_bytes": exc.limit_bytes,
                },
            },
        ) from exc
    except ProjectNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ProjectAccessError as exc:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc
